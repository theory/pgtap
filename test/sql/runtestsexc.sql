\unset ECHO
\i test/setup.sql
SET client_min_messages = warning;

CREATE SCHEMA whatever;
CREATE TABLE whatever.foo ( id serial primary key );

-- Make sure we get test function names.
SET client_min_messages = notice;

CREATE OR REPLACE FUNCTION whatever.test1this() RETURNS SETOF TEXT AS $$
    SELECT pass('simple pass') AS foo
    UNION SELECT pass('another simple pass')
    ORDER BY foo ASC;
$$ LANGUAGE SQL;

-- Make sure we have tests after the test with exception
CREATE OR REPLACE FUNCTION whatever.test2unexpectedexception() RETURNS SETOF TEXT AS $$
BEGIN
    RAISE EXCEPTION 'Runner should continue after exception in test';
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION whatever.test3plpgsql() RETURNS SETOF TEXT AS $$
BEGIN
    RETURN NEXT pass( 'plpgsql simple' );
    RETURN NEXT fail( 'simple fail' );
    RETURN NEXT pass( 'plpgsql simple 2' );
    INSERT INTO whatever.foo VALUES(1);
    RETURN NEXT is( MAX(id), 1, 'Should be a 1 in the test table') FROM whatever.foo;
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION whatever.testunexpectedexception() RETURNS SETOF TEXT AS $$
BEGIN
    RAISE EXCEPTION 'Runner should continue after exception in test';
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION whatever.teardownunexpectedexception() RETURNS SETOF TEXT AS $$
BEGIN
    RAISE EXCEPTION 'Runner should continue after exception in teardown';
END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE FUNCTION whatever.shutdownunexpectedexception() RETURNS SETOF TEXT AS $$
BEGIN
    RAISE EXCEPTION 'Runner should continue after exception in shutdown';
END;
$$ LANGUAGE plpgsql;

-- Run the actual tests. Yes, it's a one-liner!
SELECT * FROM runtests('whatever'::name);

ROLLBACK;
