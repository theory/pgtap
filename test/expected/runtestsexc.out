\unset ECHO
# whatever.test1this()
ok 1 - simple pass
ok 2 - another simple pass
not ok 3 - Unexpected exception
# Failed test 3: "Unexpected exception"
#         died: P0001: Runner should continue after exception in teardown
# whatever.test2unexpectedexception()
not ok 4 - Unexpected exception
# Failed test 4: "Unexpected exception"
#         died: P0001: Runner should continue after exception in test
# whatever.test3plpgsql()
ok 5 - plpgsql simple
not ok 6 - simple fail
# Failed test 6: "simple fail"
ok 7 - plpgsql simple 2
ok 8 - Should be a 1 in the test table
not ok 9 - Unexpected exception
# Failed test 9: "Unexpected exception"
#         died: P0001: Runner should continue after exception in teardown
# whatever.testunexpectedexception()
not ok 10 - Unexpected exception
# Failed test 10: "Unexpected exception"
#         died: P0001: Runner should continue after exception in test
not ok 11 - Unexpected exception in shutdown
# Failed test 11: "Unexpected exception in shutdown"
#         died: P0001: Runner should continue after exception in shutdown
1..11
# Looks like you failed 6 tests of 11
